text/0000-target-stages.md

```diff
@@ -0,0 +1,146 @@
+with higher target-stages can benefit from those stage dependencies being tracked.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+- Potentially buggy release artifacts, if the compiler isn't careful enough, it might reutilize an artifact that it shouldn't and result in a buggy binary without any human error by the user.
+- Coordination between teams and sub-projects is lengthy and a serious effort. I can expect that both Cargo and the compiler will be greatly involved, at least in some portion.
+- A potential for bigger `target` sizes / quicker rise in `target` size, if the user uses multiple "bases", using each one with multiple, different "data dependencies". This is only theorical and depends on the final algorithm used, as the intent is for `target` sizes to be reduced.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- The current incremental system is an alternative. But users find that unnecessary rebuilds are a problem.
+- This flag would override the currently unstable and untracked `no_analysis` option. Althought this flag currently
+exists, it doesn't see any real use inside the compiler, nor does it have a tracking issue. It doesn't operate with
+Cargo, and isn't designed to be as loadbearing as this RFC proposes.
+
+
+# Prior art
+[prior-art]: #prior-art
+
+A few examples of what this can include are:
+
+- [GCC IncrementalCompiler Project][gccincr], currently without a Delivery Date (last update; 2008).
+- Clang has the [Clang-Repl project][clang-repl], a C++ interpreter which supports incremental compilation
+- [Zig Incremental compilation][zig-incr], doesn't have a functional incremental compilation system for many use-cases.
+- [Go (incremental?) compiler][go-page], supposedly incremental but I cannot find documentation on this or an implementation. 
+- [Ocaml incremental compiler][ocaml-incr], the code is available, but not documented to an extensive degree.
+- [This Project Goal][project-goal], that talks about a similar concept, but with a more primitive approach.
+
+It seems that Rust currently has the most documented incremental compilation system.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- New design of the incremental directory.
+- How will these target stages be handled when saving a dependency graph?
+- Where will the stage dependencies be stored? Should we store them along the dependency graph?
+- How does this effort interact with the [Relink, Don't Rebuild][rdr] project goal?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+This effort helps push forward towards a query-driven compiler, where all big "stages" of a compilation are
+querified.
+
+It also allows for better flexibility when interacting with 3rd party programs, and ad-hoc programs
+to `rustc_driver` like Clippy and thus, allowing the share of cache between these.
```

Comment:
Incr comp strongly depends on the compiler source remaining unchanged. Clippy has a different compiler source from rustc and does in fact override some untracked things like the list of lints (it can't be tracked because you can't serialize a lint pass trait object).

Reply:
If 3rd party programs have the ability to alter rustc_driver that much, supporting them is outside the scope of this RFC (and I don't think that we could have an sustainable universal solution for this kind of use case.) I'll re-word it.